vault/token_store.go

```diff
@@ -1213,93 +1213,104 @@ func (ts *TokenStore) handleCreateCommon(
 
 	resp := &logical.Response{}
 
-	switch {
-	// If we have a role, and the role defines policies, we don't even consider
-	// parent policies; the role allowed policies trumps all
-	case role != nil && len(role.AllowedPolicies) > 0:
-		sanitizedRolePolicies := policyutil.SanitizePolicies(role.AllowedPolicies, false)
-
-		if len(data.Policies) == 0 {
-			data.Policies = sanitizedRolePolicies
-		} else {
-			// Sanitize passed-in and role policies before comparison
-			sanitizedInputPolicies := policyutil.SanitizePolicies(data.Policies, false)
+	var addDefault bool
 
-			if !strutil.StrListSubset(sanitizedRolePolicies, sanitizedInputPolicies) {
-				return logical.ErrorResponse(fmt.Sprintf("token policies (%v) must be subset of the role's allowed policies (%v)", sanitizedInputPolicies, sanitizedRolePolicies)), logical.ErrInvalidRequest
-			}
+	switch {
+	case role != nil && (len(role.AllowedPolicies) > 0 || len(role.DisallowedPolicies) > 0):
+		// Holds the final set of policies as they get munged
+		var finalPolicies []string
+
+		// We don't make use of the global one because roles with allowed or
+		// disallowed set do their own policy rules
+		var localAddDefault bool
+
+		// If the request doesn't say not to add "default" and if "default"
+		// isn't in the disallowed list, add it. This is in line with the idea
+		// that roles, when allowed/disallowed ar set, allow a subset of
+		// policies to be set disjoint from the parent token's policies.
+		if !data.NoDefaultPolicy && !strutil.StrListContains(role.DisallowedPolicies, "default") {
+			localAddDefault = true
 		}
 
-		// If AllowedPolicies does not have "default" policy in
-		// it, then the token being created should not have it
-		// either.
-		if !strutil.StrListContains(sanitizedRolePolicies, "default") {
-			data.NoDefaultPolicy = true
+		// Start with passed-in policies as a baseline, if they exist
+		if len(data.Policies) > 0 {
+			finalPolicies = policyutil.SanitizePolicies(data.Policies, localAddDefault)
```

Comment:
There is a catch here. Notice that for `policyutil.SanitizePolicies`, the second parameter is just a hint whether to add `default` or not, "if not already present". In case if `data.Policies` has `default` in it already, then `localAddDefault` being `true` does not ensure that `finalPolicies` is not having `default`.